receiver/k8slogreceiver/design.md

```diff
@@ -0,0 +1,120 @@
+# Overview
+
+The k8slogreceiver is a receiver that receives logs from Kubernetes pods.
+
+```
+   K8s
+   API ─────┐
+            │
+ Docker     │
+  API ────┐ │
+          │ │
+Cri       │ │
+API ────┐ │ │
+        │ │ │
+        ▼ ▼ ▼
+    ┌──────────────┐         ┌───────────┐
+    │    Source    │────────▶│   Poller  │
+    └──────────────┘         └───────────┘
+          │                       │
+          │                       │ 
+          ▼                       │
+    ┌───────────────┐             │
+    │  Reader       │◀────────────┘
+    └───────────────┘
+```
+
+The `Source` will get information about pods from the Kubernetes API and will get the paths will log files exist from the Docker API and the CRI API (Only containerd is supported now).
+
+Then, the files will be found and read by the `Poller` and `Reader`.
+
+More detailed, for the three mode, the design is as follows:
+
+## daemonset-stdout
+
+```
+ k8sapi     ┌──────────────┐    docker/cri-containerd
+   │        │              │             │
+   └───────▶│    Source    │◀────────────┘
+  medatada  │              │ logPath, env
+            └──────────────┘
+                  │ assosiate
+                  ▼
+            ┌──────────────┐
+            │   Reader     │
+            └──────────────┘
+                  │ read files from logPath
+                  ▼
+                files
+```
+
+The `Source` will get pod list and metadata from k8s api (maybe there will also be an implementation using kubelet).
+Then, it will get the logPath and env from docker/cri-containerd api. And they will be associated with the pod metadata
```

Comment:
Can't we just by-pass the extra API call to the runtime and construct the log path based on a default pattern which is more or less known? i.e. `/var/log/pods/'<namespace>_<pod_name>_<pod_uid>'/<container_name>/*.log`, where `namespace`, `pod_name`, `pod_uid` and `container_name` are already known from the k8s API.

Reply:
This also works. Datadog is using this way. But:
1. The path is different across k8s versions. Maybe it will also change in future. see https://github.com/DataDog/datadog-agent/blob/0d3dcb9fdb96e8a9d1810bb041d2dae6250372f8/pkg/logs/launchers/container/tailerfactory/file.go#L239
2. Some runtime API will directly return the actual path, which reduces resolving symbol links.
3. If we bypass extra API, it will be complex to get environment variables defined by downward api.